README.md
> [!IMPORTANT]  
> This repo has moved to https://github.com/dart-lang/test/tree/master/pkgs/test_descriptor

[![Dart CI](https://github.com/dart-lang/test_descriptor/actions/workflows/test-package.yml/badge.svg)](https://github.com/dart-lang/test_descriptor/actions/workflows/test-package.yml)
[![pub package](https://img.shields.io/pub/v/test_descriptor.svg)](https://pub.dev/packages/test_descriptor)
[![package publisher](https://img.shields.io/pub/publisher/test_descriptor.svg)](https://pub.dev/packages/test_descriptor/publisher)

The `test_descriptor` package provides a convenient, easy-to-read API for
defining and verifying directory structures in tests.

## Usage

We recommend that you import this library with the `d` prefix. The
[`d.dir()`][dir] and [`d.file()`][file] functions are the main entrypoints. They
define a filesystem structure that can be created using
[`Descriptor.create()`][create] and verified using
[`Descriptor.validate()`][validate]. For example:

[dir]: https://pub.dev/documentation/test_descriptor/latest/test_descriptor/dir.html
[file]: https://pub.dev/documentation/test_descriptor/latest/test_descriptor/file.html
[create]: https://pub.dev/documentation/test_descriptor/latest/test_descriptor/Descriptor/create.html
[validate]: https://pub.dev/documentation/test_descriptor/latest/test_descriptor/Descriptor/validate.html

```dart
import 'dart:io';

import 'package:test/test.dart';
import 'package:test_descriptor/test_descriptor.dart' as d;

void main() {
  test('Directory.rename', () async {
    await d.dir('parent', [
      d.file('sibling', 'sibling-contents'),
      d.dir('old-name', [d.file('child', 'child-contents')])
    ]).create();

    await Directory('${d.sandbox}/parent/old-name')
        .rename('${d.sandbox}/parent/new-name');

    await d.dir('parent', [
      d.file('sibling', 'sibling-contents'),
      d.dir('new-name', [d.file('child', 'child-contents')])
    ]).validate();
  });
}
```

By default, descriptors create entries in a temporary sandbox directory,
[`d.sandbox`][sandbox]. A new sandbox is automatically created the first time
you create a descriptor in a given test, and automatically deleted once the test
finishes running.

[sandbox]: https://pub.dev/documentation/test_descriptor/latest/test_descriptor/sandbox.html

This package is [`term_glyph`][term_glyph] aware. It will decide whether to use
ASCII or Unicode glyphs based on the [`glyph.ascii`][ascii] attribute.

[term_glyph]: https://pub.dev/packages/term_glyph
[ascii]: https://pub.dev/documentation/term_glyph/latest/term_glyph/ascii.html


CHANGELOG.md
## 2.0.1

* Update `meta` constraint to `>=1.3.0 <3.0.0`.

## 2.0.0

* Null safety stable release.
* See breaking changes in the `2.0.0-nullsafety` version.

## 2.0.0-nullsafety

* Migrate to null safety.
* Fix outdated URLs in `README.md`.
* BREAKING: Removed archive support.
* BREAKING: `DirectoryDescriptor.load` only supports a `String` path instead of
  also accepting relative `Uri` objects.
* BREAKING: `DirectoryDescriptor.load` no longer has an optional `parents`
  parameter - this was intended for internal use only.

## 1.2.0

* Add an `ArchiveDescriptor` class and a corresponding `archive()` function that
  can create and validate Zip and TAR archives.

## 1.1.1

* Update to lowercase Dart core library constants.

## 1.1.0

* Add a `path()` function that returns the a path within the sandbox directory.

* Add `io` getters to `FileDescriptor` and `DirectoryDescriptor` that returns
  `dart:io` `File` and `Directory` objects, respectively, within the sandbox
  directory.

## 1.0.4

* Support test `1.x.x'.

## 1.0.3

* Stop using comment-based generics.

## 1.0.2

* Declare support for `async` 2.0.0.

## 1.0.1

* `FileDescriptor.validate()` now allows invalid UTF-8 files.

* Fix a bug where `DirectoryDescriptor.load()` would incorrectly report that
  multiple versions of a file or directory existed.

## 1.0.0

* Initial version.


test/archive_test.dart
// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

@TestOn('vm')

import 'dart:convert';
import 'dart:io';

import 'package:archive/archive.dart';
import 'package:async/async.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';
import 'package:test_descriptor/test_descriptor.dart' as d;

import 'utils.dart';

void main() {
  group("create()", () {
    test("creates an empty archive", () async {
      await d.archive("test.tar").create();

      var archive =
          TarDecoder().decodeBytes(File(d.path("test.tar")).readAsBytesSync());
      expect(archive.files, isEmpty);
    });

    test("creates an archive with files", () async {
      await d.archive("test.tar", [
        d.file("file1.txt", "contents 1"),
        d.file("file2.txt", "contents 2")
      ]).create();

      var files = TarDecoder()
          .decodeBytes(File(d.path("test.tar")).readAsBytesSync())
          .files;
      expect(files.length, equals(2));
      _expectFile(files[0], "file1.txt", "contents 1");
      _expectFile(files[1], "file2.txt", "contents 2");
    });

    test("creates an archive with files in a directory", () async {
      await d.archive("test.tar", [
        d.dir("dir", [
          d.file("file1.txt", "contents 1"),
          d.file("file2.txt", "contents 2")
        ])
      ]).create();

      var files = TarDecoder()
          .decodeBytes(File(d.path("test.tar")).readAsBytesSync())
          .files;
      expect(files.length, equals(2));
      _expectFile(files[0], "dir/file1.txt", "contents 1");
      _expectFile(files[1], "dir/file2.txt", "contents 2");
    });

    test("creates an archive with files in a nested directory", () async {
      await d.archive("test.tar", [
        d.dir("dir", [
          d.dir("subdir", [
            d.file("file1.txt", "contents 1"),
            d.file("file2.txt", "contents 2")
          ])
        ])
      ]).create();

      var files = TarDecoder()
          .decodeBytes(File(d.path("test.tar")).readAsBytesSync())
          .files;
      expect(files.length, equals(2));
      _expectFile(files[0], "dir/subdir/file1.txt", "contents 1");
      _expectFile(files[1], "dir/subdir/file2.txt", "contents 2");
    });

    group("creates a file in", () {
      test("zip format", () async {
        await d.archive("test.zip", [d.file("file.txt", "contents")]).create();

        var archive = ZipDecoder()
            .decodeBytes(File(d.path("test.zip")).readAsBytesSync());
        _expectFile(archive.files.single, "file.txt", "contents");
      });

      group("gzip tar format", () {
        for (var extension in [".tar.gz", ".tar.gzip", ".tgz"]) {
          test("with $extension", () async {
            await d.archive(
                "test$extension", [d.file("file.txt", "contents")]).create();

            var archive = TarDecoder().decodeBytes(GZipDecoder()
                .decodeBytes(File(d.path("test$extension")).readAsBytesSync()));
            _expectFile(archive.files.single, "file.txt", "contents");
          });
        }
      });

      group("bzip2 tar format", () {
        for (var extension in [".tar.bz2", ".tar.bzip2"]) {
          test("with $extension", () async {
            await d.archive(
                "test$extension", [d.file("file.txt", "contents")]).create();

            var archive = TarDecoder().decodeBytes(BZip2Decoder()
                .decodeBytes(File(d.path("test$extension")).readAsBytesSync()));
            _expectFile(archive.files.single, "file.txt", "contents");
          });
        }
      });
    });

    group("gracefully rejects", () {
      test("an uncreatable descriptor", () async {
        await expectLater(
            d.archive("test.tar", [d.filePattern(RegExp(r"^foo-"))]).create(),
            throwsUnsupportedError);
        await d.nothing("test.tar").validate();
      });

      test("a non-file non-directory descriptor", () async {
        await expectLater(
            d.archive("test.tar", [d.nothing("file.txt")]).create(),
            throwsUnsupportedError);
        await d.nothing("test.tar").validate();
      });

      test("an unknown file extension", () async {
        await expectLater(
            d.archive("test.asdf", [d.nothing("file.txt")]).create(),
            throwsUnsupportedError);
      });
    });
  });

  group("validate()", () {
    group("with an empty archive", () {
      test("succeeds if an empty archive exists", () async {
        File(d.path("test.tar"))
            .writeAsBytesSync(TarEncoder().encode(Archive()));
        await d.archive("test.tar").validate();
      });

      test("succeeds if a non-empty archive exists", () async {
        File(d.path("test.tar")).writeAsBytesSync(
            TarEncoder().encode(Archive()..addFile(_file("file.txt"))));
        await d.archive("test.tar").validate();
      });

      test("fails if no archive exists", () {
        expect(d.archive("test.tar").validate(),
            throwsA(toString(startsWith('File not found: "test.tar".'))));
      });

      test("fails if an invalid archive exists", () {
        d.file("test.tar", "not a valid tar file").create();
        expect(
            d.archive("test.tar").validate(),
            throwsA(toString(
                startsWith('File "test.tar" is not a valid archive.'))));
      });
    });

    test("succeeds if an archive contains a matching file", () async {
      File(d.path("test.tar")).writeAsBytesSync(TarEncoder()
          .encode(Archive()..addFile(_file("file.txt", "contents"))));
      await d.archive("test.tar", [d.file("file.txt", "contents")]).validate();
    });

    test("fails if an archive doesn't contain a file", () async {
      File(d.path("test.tar")).writeAsBytesSync(TarEncoder().encode(Archive()));
      expect(
          d.archive("test.tar", [d.file("file.txt", "contents")]).validate(),
          throwsA(
              toString(startsWith('File not found: "test.tar/file.txt".'))));
    });

    test("fails if an archive contains a non-matching file", () async {
      File(d.path("test.tar")).writeAsBytesSync(TarEncoder()
          .encode(Archive()..addFile(_file("file.txt", "wrong contents"))));
      expect(
          d.archive("test.tar", [d.file("file.txt", "contents")]).validate(),
          throwsA(toString(
              startsWith('File "test.tar/file.txt" should contain:'))));
    });

    test("succeeds if an archive contains a file matching a pattern", () async {
      File(d.path("test.tar")).writeAsBytesSync(TarEncoder()
          .encode(Archive()..addFile(_file("file.txt", "contents"))));
      await d.archive("test.tar",
          [d.filePattern(RegExp(r"f..e\.txt"), "contents")]).validate();
    });

    group("validates a file in", () {
      test("zip format", () async {
        File(d.path("test.zip")).writeAsBytesSync(ZipEncoder()
            .encode(Archive()..addFile(_file("file.txt", "contents"))));

        await d
            .archive("test.zip", [d.file("file.txt", "contents")]).validate();
      });

      group("gzip tar format", () {
        for (var extension in [".tar.gz", ".tar.gzip", ".tgz"]) {
          test("with $extension", () async {
            File(d.path("test$extension")).writeAsBytesSync(GZipEncoder()
                .encode(TarEncoder().encode(
                    Archive()..addFile(_file("file.txt", "contents")))));

            await d.archive(
                "test$extension", [d.file("file.txt", "contents")]).validate();
          });
        }
      });

      group("bzip2 tar format", () {
        for (var extension in [".tar.bz2", ".tar.bzip2"]) {
          test("with $extension", () async {
            File(d.path("test$extension")).writeAsBytesSync(BZip2Encoder()
                .encode(TarEncoder().encode(
                    Archive()..addFile(_file("file.txt", "contents")))));

            await d.archive(
                "test$extension", [d.file("file.txt", "contents")]).validate();
          });
        }
      });
    });

    test("gracefully rejects an unknown file format", () {
      expect(d.archive("test.asdf").validate(), throwsUnsupportedError);
    });
  });

  test("read() is unsupported", () {
    expect(d.archive("test.tar").read(), throwsUnsupportedError);
  });

  test("readAsBytes() returns the contents of the archive", () async {
    var descriptor = d.archive("test.tar",
        [d.file("file1.txt", "contents 1"), d.file("file2.txt", "contents 2")]);

    var files = TarDecoder()
        .decodeBytes(await collectBytes(descriptor.readAsBytes()))
        .files;
    expect(files.length, equals(2));
    _expectFile(files[0], "file1.txt", "contents 1");
    _expectFile(files[1], "file2.txt", "contents 2");
  });

  test("archive returns the in-memory contents", () async {
    var archive = await d
        .archive("test.tar", [
          d.file("file1.txt", "contents 1"),
          d.file("file2.txt", "contents 2")
        ])
        .archive;

    var files = archive.files;
    expect(files.length, equals(2));
    _expectFile(files[0], "file1.txt", "contents 1");
    _expectFile(files[1], "file2.txt", "contents 2");
  });

  test("io refers to the file within the sandbox", () {
    expect(d.file('test.tar').io.path, equals(p.join(d.sandbox, 'test.tar')));
  });
}

/// Asserts that [file] has the given [name] and [contents].
void _expectFile(ArchiveFile file, String name, String contents) {
  expect(file.name, equals(name));
  expect(utf8.decode(file.content as List<int>), equals(contents));
}

/// Creates an [ArchiveFile] with the given [name] and [contents].
ArchiveFile _file(String name, [String contents]) {
  var bytes = utf8.encode(contents ?? "");
  return ArchiveFile(name, bytes.length, bytes)
    // Setting the mode and mod time are necessary to work around
    // brendan-duncan/archive#76.
    ..mode = 428
    ..lastModTime = DateTime.now().millisecondsSinceEpoch ~/ 1000;
}


lib/test_descriptor.dart
// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:path/path.dart' as p;
import 'package:test/test.dart';

import 'src/descriptor.dart';
import 'src/directory_descriptor.dart';
import 'src/file_descriptor.dart';
import 'src/nothing_descriptor.dart';
import 'src/pattern_descriptor.dart';
import 'src/sandbox.dart';

export 'src/descriptor.dart';
export 'src/directory_descriptor.dart';
export 'src/file_descriptor.dart';
export 'src/nothing_descriptor.dart';
export 'src/pattern_descriptor.dart';
export 'src/sandbox.dart' show sandbox;

/// Creates a new [FileDescriptor] with [name] and [contents].
///
/// The [contents] may be a `String`, a `List<int>`, or a [Matcher]. If it's a
/// string, [Descriptor.create] creates a UTF-8 file and [Descriptor.validate]
/// parses the physical file as UTF-8. If it's a [Matcher],
/// [Descriptor.validate] matches it against the physical file's contents parsed
/// as UTF-8, and [Descriptor.create] is unsupported.
///
/// If [contents] isn't passed, [Descriptor.create] creates an empty file and
/// [Descriptor.validate] verifies that the file is empty.
///
/// To match a [Matcher] against a file's binary contents, use [new
/// FileDescriptor.binaryMatcher] instead.
FileDescriptor file(String name, [contents]) => FileDescriptor(name, contents);

/// Creates a new [DirectoryDescriptor] descriptor with [name] and [contents].
///
/// [Descriptor.validate] requires that all descriptors in [contents] match
/// children of the physical diretory, but it *doesn't* require that no other
/// children exist. To ensure that a particular child doesn't exist, use
/// [nothing].
DirectoryDescriptor dir(String name, [Iterable<Descriptor> contents]) =>
    DirectoryDescriptor(name, contents ?? <Descriptor>[]);

/// Creates a new [NothingDescriptor] descriptor that asserts that no entry
/// named [name] exists.
///
/// [Descriptor.create] does nothing for this descriptor.
NothingDescriptor nothing(String name) => NothingDescriptor(name);

/// Creates a new [PatternDescriptor] descriptor that asserts than an entry with
/// a name matching [pattern] exists, and matches the [Descriptor] returned
/// by [child].
///
/// The [child] callback is passed the basename of each entry matching [name].
/// It returns a descriptor that should match that entry. It's valid for
/// multiple entries to match [name] as long as only one of them matches
/// [child].
///
/// [Descriptor.create] is not supported for this descriptor.
PatternDescriptor pattern(
        Pattern name, Descriptor Function(String basename) child) =>
    PatternDescriptor(name, child);

/// A convenience method for creating a [PatternDescriptor] descriptor that
/// constructs a [FileDescriptor] descriptor.
PatternDescriptor filePattern(Pattern name, [contents]) =>
    pattern(name, (realName) => file(realName, contents));

/// A convenience method for creating a [PatternDescriptor] descriptor that
/// constructs a [DirectoryDescriptor] descriptor.
PatternDescriptor dirPattern(Pattern name, [Iterable<Descriptor> contents]) =>
    pattern(name, (realName) => dir(realName, contents));

/// Returns [path] within the [sandbox] directory.
String path(String path) => p.join(sandbox, path);


lib/src/file_descriptor.dart
// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:async';
import 'dart:io';
import 'dart:math' as math;

import 'package:collection/collection.dart';
import 'package:path/path.dart' as p;
import 'package:term_glyph/term_glyph.dart' as glyph;
import 'package:test/test.dart';

import 'descriptor.dart';
import 'sandbox.dart';
import 'utils.dart';

/// A descriptor describing a single file.
///
/// In addition to the normal descriptor methods, this has [read] and
/// [readAsBytes] methods that allows its contents to be read.
///
/// This may be extended outside this package.
abstract class FileDescriptor extends Descriptor {
  /// Creates a new [FileDescriptor] with [name] and [contents].
  ///
  /// The [contents] may be a `String`, a `List<int>`, or a [Matcher]. If it's a
  /// string, [create] creates a UTF-8 file and [validate] parses the physical
  /// file as UTF-8. If it's a [Matcher], [validate] matches it against the
  /// physical file's contents parsed as UTF-8, and [create], [read], and
  /// [readAsBytes] are unsupported.
  ///
  /// If [contents] isn't passed, [create] creates an empty file and [validate]
  /// verifies that the file is empty.
  ///
  /// To match a [Matcher] against a file's binary contents, use [new
  /// FileDescriptor.binaryMatcher] instead.
  factory FileDescriptor(String name, contents) {
    if (contents is String) return _StringFileDescriptor(name, contents);
    if (contents is List) {
      return _BinaryFileDescriptor(name, contents.cast<int>());
    }
    if (contents == null) return _BinaryFileDescriptor(name, []);
    return _MatcherFileDescriptor(name, contents as Matcher);
  }

  /// Returns a `dart:io` [File] object that refers to this file within
  /// [sandbox].
  File get io => File(p.join(sandbox, name));

  /// Creates a new binary [FileDescriptor] with [name] that matches its binary
  /// contents against [matcher].
  ///
  /// The [create], [read], and [readAsBytes] methods are unsupported for this
  /// descriptor.
  factory FileDescriptor.binaryMatcher(String name, Matcher matcher) =>
      _MatcherFileDescriptor(name, matcher, isBinary: true);

  /// A protected constructor that's only intended for subclasses.
  FileDescriptor.protected(String name) : super(name);

  @override
  Future<void> create([String? parent]) async {
    // Create the stream before we call [File.openWrite] because it may fail
    // fast (e.g. if this is a matcher file).
    var file = File(p.join(parent ?? sandbox, name)).openWrite();
    try {
      await readAsBytes().listen(file.add).asFuture();
    } finally {
      await file.close();
    }
  }

  @override
  Future<void> validate([String? parent]) async {
    var fullPath = p.join(parent ?? sandbox, name);
    var pretty = prettyPath(fullPath);
    if (!(await File(fullPath).exists())) {
      fail('File not found: "$pretty".');
    }

    await _validate(pretty, await File(fullPath).readAsBytes());
  }

  /// Validates that [binaryContents] matches the expected contents of
  /// the descriptor.
  ///
  /// The [prettyPath] is a human-friendly representation of the path to the
  /// descriptor.
  FutureOr<void> _validate(String prettyPath, List<int> binaryContents);

  /// Reads and decodes the contents of this descriptor as a UTF-8 string.
  ///
  /// This isn't supported for matcher descriptors.
  Future<String> read() => utf8.decodeStream(readAsBytes());

  /// Reads the contents of this descriptor as a byte stream.
  ///
  /// This isn't supported for matcher descriptors.
  Stream<List<int>> readAsBytes();

  @override
  String describe() => name;
}

class _BinaryFileDescriptor extends FileDescriptor {
  /// The contents of this descriptor's file.
  final List<int> _contents;

  _BinaryFileDescriptor(String name, this._contents) : super.protected(name);

  @override
  Stream<List<int>> readAsBytes() => Stream.fromIterable([_contents]);

  @override
  Future<void> _validate(String prettPath, List<int> actualContents) async {
    if (const IterableEquality().equals(_contents, actualContents)) return null;
    // TODO(nweiz): show a hex dump here if the data is small enough.
    fail('File "$prettPath" didn\'t contain the expected binary data.');
  }
}

class _StringFileDescriptor extends FileDescriptor {
  /// The contents of this descriptor's file.
  final String _contents;

  _StringFileDescriptor(String name, this._contents) : super.protected(name);

  @override
  Future<String> read() async => _contents;

  @override
  Stream<List<int>> readAsBytes() =>
      Stream.fromIterable([utf8.encode(_contents)]);

  @override
  void _validate(String prettyPath, List<int> actualContents) {
    var actualContentsText = utf8.decode(actualContents);
    if (_contents == actualContentsText) return null;
    fail(_textMismatchMessage(prettyPath, _contents, actualContentsText));
  }

  String _textMismatchMessage(
      String prettyPath, String expected, String actual) {
    final expectedLines = expected.split('\n');
    final actualLines = actual.split('\n');

    var results = [];

    // Compare them line by line to see which ones match.
    var length = math.max(expectedLines.length, actualLines.length);
    for (var i = 0; i < length; i++) {
      if (i >= actualLines.length) {
        // Missing output.
        results.add('? ${expectedLines[i]}');
      } else if (i >= expectedLines.length) {
        // Unexpected extra output.
        results.add('X ${actualLines[i]}');
      } else {
        var expectedLine = expectedLines[i];
        var actualLine = actualLines[i];

        if (expectedLine != actualLine) {
          // Mismatched lines.
          results.add('X $actualLine');
        } else {
          // Matched lines.
          results.add('${glyph.verticalLine} $actualLine');
        }
      }
    }

    return 'File "$prettyPath" should contain:\n'
        '${addBar(expected)}\n'
        'but actually contained:\n'
        "${results.join('\n')}";
  }
}

class _MatcherFileDescriptor extends FileDescriptor {
  /// The matcher for this descriptor's contents.
  final Matcher _matcher;

  /// Whether [_matcher] should match against the file's string or byte
  /// contents.
  final bool _isBinary;

  _MatcherFileDescriptor(String name, this._matcher, {bool isBinary = false})
      : _isBinary = isBinary,
        super.protected(name);

  @override
  Stream<List<int>> readAsBytes() =>
      throw UnsupportedError("Matcher files can't be created or read.");

  @override
  Future<void> _validate(String prettyPath, List<int> actualContents) async {
    try {
      expect(
          _isBinary ? actualContents : utf8.decode(actualContents), _matcher);
    } on TestFailure catch (error) {
      fail('Invalid contents for file "$prettyPath":\n${error.message}');
    }
  }
}


lib/src/directory_descriptor.dart
// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:async';
import 'dart:io';

import 'package:async/async.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

import 'descriptor.dart';
import 'file_descriptor.dart';
import 'sandbox.dart';
import 'utils.dart';

/// A descriptor describing a directory that may contain nested descriptors.
///
/// In addition to the normal descriptor methods, this has a [load] method that
/// allows it to be used as a virtual filesystem.
///
/// This may be extended outside this package.
class DirectoryDescriptor extends Descriptor {
  /// Descriptors for entries in this directory.
  ///
  /// This may be modified.
  final List<Descriptor> contents;

  /// Returns a `dart:io` [Directory] object that refers to this file within
  /// [sandbox].
  Directory get io => Directory(p.join(sandbox, name));

  DirectoryDescriptor(String name, Iterable<Descriptor> contents)
      : contents = contents.toList(),
        super(name);

  /// Creates a directory descriptor named [name] that describes the physical
  /// directory at [path].
  factory DirectoryDescriptor.fromFilesystem(String name, String path) {
    return DirectoryDescriptor(
        name,
        Directory(path).listSync().map((entity) {
          // Ignore hidden files.
          if (p.basename(entity.path).startsWith('.')) return null;

          if (entity is Directory) {
            return DirectoryDescriptor.fromFilesystem(
                p.basename(entity.path), entity.path);
          } else if (entity is File) {
            return FileDescriptor(
                p.basename(entity.path), entity.readAsBytesSync());
          }
          // Ignore broken symlinks.
          return null;
        }).whereType<Descriptor>());
  }

  @override
  Future<void> create([String? parent]) async {
    var fullPath = p.join(parent ?? sandbox, name);
    await Directory(fullPath).create(recursive: true);
    await Future.wait(contents.map((entry) => entry.create(fullPath)));
  }

  @override
  Future<void> validate([String? parent]) async {
    var fullPath = p.join(parent ?? sandbox, name);
    if (!(await Directory(fullPath).exists())) {
      fail('Directory not found: "${prettyPath(fullPath)}".');
    }

    await waitAndReportErrors(
        contents.map((entry) => entry.validate(fullPath)));
  }

  /// Treats this descriptor as a virtual filesystem and loads the binary
  /// contents of the [FileDescriptor] at the given relative [path].
  ///
  /// The [parents] parameter should only be passed by subclasses of
  /// [DirectoryDescriptor] that are recursively calling [load]. It's the
  /// URL-format path of the directories that have been loaded so far.
  Stream<List<int>> load(String path, [String? parents]) {
    if (!p.url.isWithin('.', path)) {
      throw ArgumentError.value(
          path, 'path', 'must be relative and beneath the base URL.');
    }

    return StreamCompleter.fromFuture(Future.sync(() {
      var split = p.url.split(p.url.normalize(path));
      var file = split.length == 1;
      var matchingEntries = contents.where((entry) {
        return entry.name == split.first &&
            (file ? entry is FileDescriptor : entry is DirectoryDescriptor);
      }).toList();

      var type = file ? 'file' : 'directory';
      var parentsAndSelf = parents == null ? name : p.url.join(parents, name);
      if (matchingEntries.isEmpty) {
        fail('Couldn\'t find a $type descriptor named "${split.first}" within '
            '"$parentsAndSelf".');
      } else if (matchingEntries.length > 1) {
        fail('Found multiple $type descriptors named "${split.first}" within '
            '"$parentsAndSelf".');
      } else {
        var remainingPath = split.sublist(1);
        if (remainingPath.isEmpty) {
          return (matchingEntries.first as FileDescriptor).readAsBytes();
        } else {
          return (matchingEntries.first as DirectoryDescriptor)
              .load(p.url.joinAll(remainingPath), parentsAndSelf);
        }
      }
    }));
  }

  @override
  String describe() => describeDirectory(name, contents);
}


lib/src/archive_descriptor.dart
// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:async';
import 'dart:io';

import 'package:archive/archive.dart';
import 'package:async/async.dart';
import 'package:meta/meta.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

import 'descriptor.dart';
import 'directory_descriptor.dart';
import 'file_descriptor.dart';
import 'sandbox.dart';
import 'utils.dart';

/// A [Descriptor] describing files in a Tar or Zip archive.
///
/// The format is determined by the descriptor's file extension.
@sealed
class ArchiveDescriptor extends Descriptor implements FileDescriptor {
  /// Descriptors for entries in this archive.
  final List<Descriptor> contents;

  /// Returns a `package:archive` [Archive] object that contains the contents of
  /// this file.
  Future<Archive> get archive async {
    var archive = Archive();
    (await _files(contents)).forEach(archive.addFile);
    return archive;
  }

  File get io => File(p.join(sandbox, name));

  /// Returns [ArchiveFile]s for each file in [descriptors].
  ///
  /// If [parent] is passed, it's used as the parent directory for filenames.
  Future<Iterable<ArchiveFile>> _files(Iterable<Descriptor> descriptors,
      [String parent]) async {
    return (await waitAndReportErrors(descriptors.map((descriptor) async {
      var fullName =
          parent == null ? descriptor.name : "$parent/${descriptor.name}";

      if (descriptor is FileDescriptor) {
        var bytes = await collectBytes(descriptor.readAsBytes());
        return [
          ArchiveFile(fullName, bytes.length, bytes)
            // Setting the mode and mod time are necessary to work around
            // brendan-duncan/archive#76.
            ..mode = 428
            ..lastModTime = DateTime.now().millisecondsSinceEpoch ~/ 1000
        ];
      } else if (descriptor is DirectoryDescriptor) {
        return await _files(descriptor.contents, fullName);
      } else {
        throw UnsupportedError(
            "An archive can only be created from FileDescriptors and "
            "DirectoryDescriptors.");
      }
    })))
        .expand((files) => files);
  }

  ArchiveDescriptor(String name, Iterable<Descriptor> contents)
      : contents = List.unmodifiable(contents),
        super(name);

  Future create([String parent]) async {
    var path = p.join(parent ?? sandbox, name);
    var file = File(path).openWrite();
    try {
      try {
        await readAsBytes().listen(file.add).asFuture();
      } finally {
        await file.close();
      }
    } catch (_) {
      await File(path).delete();
      rethrow;
    }
  }

  Future<String> read() async => throw UnsupportedError(
      "ArchiveDescriptor.read() is not supported. Use Archive.readAsBytes() "
      "instead.");

  Stream<List<int>> readAsBytes() => Stream.fromFuture(() async {
        return _encodeFunction()(await archive);
      }());

  Future<void> validate([String parent]) async {
    // Access this first so we eaerly throw an error for a path with an invalid
    // extension.
    var decoder = _decodeFunction();

    var fullPath = p.join(parent ?? sandbox, name);
    var pretty = prettyPath(fullPath);
    if (!(await File(fullPath).exists())) {
      fail('File not found: "$pretty".');
    }

    var bytes = await File(fullPath).readAsBytes();
    Archive archive;
    try {
      archive = decoder(bytes);
    } catch (_) {
      // Catch every error to work around brendan-duncan/archive#77.
      fail('File "$pretty" is not a valid archive.');
    }

    // Because validators expect to validate against a real filesystem, we have
    // to extract the archive to a temp directory and run validation on that.
    var tempDir = await Directory.systemTemp
        .createTempSync('dart_test_')
        .resolveSymbolicLinks();

    try {
      await waitAndReportErrors(archive.files.map((file) async {
        var path = p.join(tempDir, file.name);
        await Directory(p.dirname(path)).create(recursive: true);
        await File(path).writeAsBytes(file.content as List<int>);
      }));

      await waitAndReportErrors(contents.map((entry) async {
        try {
          await entry.validate(tempDir);
        } on TestFailure catch (error) {
          // Replace the temporary directory with the path to the archive to
          // make the error more user-friendly.
          throw TestFailure(error.message.replaceAll(tempDir, pretty));
        }
      }));
    } finally {
      await Directory(tempDir).delete(recursive: true);
    }
  }

  /// Returns the function to use to encode this file to binary, based on its
  /// [name].
  List<int> Function(Archive) _encodeFunction() {
    if (name.endsWith(".zip")) {
      return ZipEncoder().encode;
    } else if (name.endsWith(".tar")) {
      return TarEncoder().encode;
    } else if (name.endsWith(".tar.gz") ||
        name.endsWith(".tar.gzip") ||
        name.endsWith(".tgz")) {
      return (archive) => GZipEncoder().encode(TarEncoder().encode(archive));
    } else if (name.endsWith(".tar.bz2") || name.endsWith(".tar.bzip2")) {
      return (archive) => BZip2Encoder().encode(TarEncoder().encode(archive));
    } else {
      throw UnsupportedError("Unknown file format $name.");
    }
  }

  /// Returns the function to use to decode this file from binary, based on its
  /// [name].
  Archive Function(List<int>) _decodeFunction() {
    if (name.endsWith(".zip")) {
      return ZipDecoder().decodeBytes;
    } else if (name.endsWith(".tar")) {
      return TarDecoder().decodeBytes;
    } else if (name.endsWith(".tar.gz") ||
        name.endsWith(".tar.gzip") ||
        name.endsWith(".tgz")) {
      return (archive) =>
          TarDecoder().decodeBytes(GZipDecoder().decodeBytes(archive));
    } else if (name.endsWith(".tar.bz2") || name.endsWith(".tar.bzip2")) {
      return (archive) =>
          TarDecoder().decodeBytes(BZip2Decoder().decodeBytes(archive));
    } else {
      throw UnsupportedError("Unknown file format $name.");
    }
  }

  String describe() => describeDirectory(name, contents);
}
